controller/api/destination/syncronized_get_stream.go

```diff
@@ -0,0 +1,74 @@
+package destination
+
+import (
+	"context"
+
+	pb "github.com/linkerd/linkerd2-proxy-api/go/destination"
+	logging "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/metadata"
+)
+
+// synchronizedGetStream is a wrapper around a pb.Destination_GetServer that
+// makes Send safe to call from multiple goroutines. It does this by using an
+// unbuffered channel to synchronize calls to Send. Since this channel is
+// unbuffered, calls to Send may block and callers should do their own queuing.
+// This type implemenets the pb.Destination_GetServer interface but only the
+// Send method is supported. Calls to other methods will panic.
+type synchronizedGetStream struct {
+	done  bool
+	ch    chan *pb.Update
+	inner pb.Destination_GetServer
+	log   *logging.Entry
+}
+
+func newSyncronizedGetStream(stream pb.Destination_GetServer, log *logging.Entry) *synchronizedGetStream {
+	return &synchronizedGetStream{
+		done:  false,
+		ch:    make(chan *pb.Update),
+		inner: stream,
+		log:   log,
+	}
+}
+
+func (s *synchronizedGetStream) SetHeader(metadata.MD) error {
+	panic("SetHeader called on synchronizedGetStream")
+}
+func (s *synchronizedGetStream) SendHeader(metadata.MD) error {
+	panic("SendHeader called on synchronizedGetStream")
+}
+func (s *synchronizedGetStream) SetTrailer(metadata.MD) {
+	panic("SetTrailer called on synchronizedGetStream")
+}
+func (s synchronizedGetStream) Context() context.Context {
+	panic("Conext called on synchronizedGetStream")
+}
+func (s *synchronizedGetStream) SendMsg(m any) error {
+	panic("SendMsg called on synchronizedGetStream")
+}
+func (s *synchronizedGetStream) RecvMsg(m any) error {
+	panic("RecvMsg called on synchronizedGetStream")
+}
+
+func (s *synchronizedGetStream) Send(update *pb.Update) error {
+	s.ch <- update
+	return nil
+}
+
+func (s *synchronizedGetStream) Start() {
+	go func() {
+		for {
+			if s.done {
+				return
+			}
+			update := <-s.ch
+			err := s.inner.Send(update)
+			if err != nil {
+				s.log.Errorf("Error sending update: %v", err)
+			}
+		}
+	}()
+}
+
+func (s *synchronizedGetStream) Stop() {
```

Comment:
I think there are two problems with this: 

1. You are modifying a value in one go-routing and reading it in another. This value is not an atomic and is not guarded by a mutex so this is potentially racey, no ? 
2. Simply flipping the bool will not always work. Imagine you are blocked on waiting for `update := <-s.ch` and you want to stop the stream. You call stop but a value is never received through the update chan. You will hang forever, no ? 

My advise is to implement the standard mechanism of supplying a closed chan and selecting over both the updates and the closed chan. This will ensure that things are thread-safe. 

```
for {
 select {
   case <-stopChan:
	 // we are done
	 return
   case update := <-s.ch:
	 // send update
   }
}
```